Unify authorization, at least partially

Arnei opened the pull request:
Helps with #5692.

Attempt at making more modules use the `XACMLAuthorizationService` (our only implementation of `AuthorizationService`) for their authorizations checks, instead of using their own custom code.

This is likely missing a few modules, most notably the AssetManager which has woven permission checking into its query structure.

### How to test this patch

There are probably many ways to test this patch. Create yourself a user with restricted roles and focus in particular on series, workflows and the search service.


### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [ ] explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch

Arnei committed:
Remove "isAuthorized" in favor of "hasPermission"

Removes the isAuthorized function from AccessControlUtil.java because it
is mostly duplicate of hasPermission from XACMLAuthorizationService.java.

If services need to check authorization, they should call an auth service
for that, and not dodge it with some static function.

Arnei committed:
Remove "isAuthorized" in favor of "hasPermission"

Removes the isAuthorized function from AccessControlUtil.java because it
is mostly duplicate of hasPermission from XACMLAuthorizationService.java.

If services need to check authorization, they should call an auth service
for that, and not dodge it with some static function.

Arnei committed:
Trying to simply auth check in WorkflowServiceImpl.java

The authorization service should check for admin roles, no need to
do it again the workflow service.

This also introduces the problem that we may only want to grant access
to an entity if it belongs to same organization as the user, no matter
what the ACL might say.

Arnei committed:
Trying to simply auth check in WorkflowServiceImpl.java

The authorization service should check for admin roles, no need to
do it again the workflow service.

This also introduces the problem that we may only want to grant access
to an entity if it belongs to same organization as the user, no matter
what the ACL might say.

Arnei committed:
Fix up javadoc in the AuthorizationService.java

Arnei committed:
Fix up javadoc in the AuthorizationService.java

Arnei committed:
Fix test

Arnei committed:
Fix test

Arnei committed:
Fix javadoc in AuthorizationService.java

Arnei committed:
Fix javadoc in AuthorizationService.java

gregorydlogan commented:
If you want this to move forward, can you rebase it against current `develop`?

KatrinIhler commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -1150,10 +1149,9 @@ protected void assertPermission(WorkflowInstance workflow, String action, String
-            || (authorizationService.hasPermission(mediapackage, action) && currentOrgId.equals(workflowOrgId));
+    boolean authorized = (currentUser.equals(workflowCreator))
+            || (authorizationService.hasPermission(
+                mediapackage, currentUser, currentOrg, action, currentOrgId.equals(workflowOrgId)));
```
The authorization service also knows the current org, so it would be enough to hand over the entity org, and have the authorization service make the check if that is not null or something.

Arnei replied:
Sure, can do that. Should be changed now.

KatrinIhler commented on modules/search-service-impl/src/main/java/org/opencastproject/search/impl/persistence/SearchServiceDatabaseImpl.java:
```diff
@@ -395,20 +406,16 @@ public MediaPackage getMediaPackage(String mediaPackageId)
+        User currentUser = securityService.getUser();
+        // There are several reasons a user may need to load a episode: to read content, to edit it, or add content
+        if (!authorizationService.hasPermission(mediaPackage, READ.toString())
+                && !authorizationService.hasPermission(mediaPackage, CONTRIBUTE.toString())
```
TIL we have an action called contribute. :D We should get rid of that one day.

KatrinIhler commented on modules/search-service-impl/src/main/java/org/opencastproject/search/impl/persistence/SearchServiceDatabaseImpl.java:
```diff
@@ -169,12 +183,10 @@ public void deleteMediaPackage(String mediaPackageId, Date deletionDate) throws
-          AccessControlList acl = AccessControlParser.parseAcl(accessControlXml);
-          if (!AccessControlUtil.isAuthorized(acl, currentUser, currentOrg, WRITE.toString(), mediaPackageId)) {
+        if (!(searchMp.isLive() && currentUser.hasRole(GLOBAL_CAPTURE_AGENT_ROLE))) {
+          if (!authorizationService.hasPermission(searchMp, currentUser, currentOrg, WRITE.toString())) {
```
For this service, you switch the permission check over from looking at the ACL directly to handing over the media package. This impacts performance a bit, because we then read the ACL from disk instead of using the one that's in the database that we already have. There's also a Github Actions failure (workflow fails during publish-engage) where I wonder if that's related. Can you take another look there?

(And I know that the permission check on the ACL directly doesn't consider the episode ID roles bc we don't know the media package id but can't we hand that over?)

Arnei replied:
Thanks for pointing that out, I was not aware that that would screw with performance. I added another `hasPermission` variants for acls that also takes a mediapackageId specifically for the episode ID roles check.

KatrinIhler commented on modules/search-service-impl/src/main/java/org/opencastproject/search/impl/SearchServiceIndex.java:
```diff
@@ -383,8 +382,7 @@ private void checkSearchEntityWritePermission(final String mediaPackageId) throw
       AccessControlList acl = persistence.getAccessControlList(mediaPackageId);
-      if (!AccessControlUtil.isAuthorized(acl, user, securityService.getOrganization(), WRITE.toString(),
-          mediaPackageId)) {
+      if (!authorizationService.hasPermission(acl, user, securityService.getOrganization(), WRITE.toString())) {
```
This probably breaks the episode id roles for this check.

Arnei replied:
Should not be broken anymore.

KatrinIhler reviewed (changes requested):
Code looks good in general, but there are some concerns left. :)

Arnei committed:
Add another hasPermission variant

This variant takes acl, but also a mediapackageId to avoid having to
get the acl from the mediapackage. This is supposed to improve
performance.

Should also fix a case were episode role id would have been ignored.

Arnei committed:
Add another hasPermission variant

This variant takes acl, but also a mediapackageId to avoid having to
get the acl from the mediapackage. This is supposed to improve
performance.

Should also fix a case were episode role id would have been ignored.

Arnei committed:
Move "boolean entityBelongsToOrg" into method

Instead of having to compute the value for entityBelongsToOrg yourself,
you now pass in the orgId of the org you want to match against the current
id and the method will check that for you.

Arnei committed:
Move "boolean entityBelongsToOrg" into method

Instead of having to compute the value for entityBelongsToOrg yourself,
you now pass in the orgId of the org you want to match against the current
id and the method will check that for you.

Arnei commented:
Not sure why the test workflow keeps failing. Can't reproduce that locally and I don't see a link between the timeouts and the PR changes.

gregorydlogan commented:
Hrm...  I can't reproduce this locally either, but the workflows are *failing*.  This, to me, suggests that one of the components supporting a WOH is broken.  Looking at the GHA logs, it's installing the system's `ffmpeg`, not `ffmpeg-dist` or anything from our repo.  Could that be the culprit?

gregorydlogan commented:
Had a dive into this.  There CI tests are reproducable locally, though I'm not exactly sure why.  The workflow is failing because the publish WOH is timing out.  Looking at the stack, I get
```
2026-08-07T15:08:49,761 | INFO  | (PublishEngageWorkflowOperationHandler:431) - Publishing media package test to search index
2026-08-07T15:13:50,924 | ERROR | (DublinCoreUtil:78) - Unable to load metadata from catalog 'http://localhost:8081/static/mh_default_org/enga
ge-player/test/0f217309-db7f-42c7-9526-739abea1b293/dublincore.xml'
org.opencastproject.util.NotFoundException: Could not copy http://localhost:8081/static/mh_default_org/engage-player/test/0f217309-db7f-42c7-9
526-739abea1b293/dublincore.xml to /home/greg/opencast/upstream/build/opencast-dist-admin/data/opencast/workspace/http_localhost_8081/static/m
h_default_org/engage-player/test/0f217309-db7f-42c7-9526-739abea1b293/dublincore-18ebd8cb-0db5-46ff-b7c4-f6f97ca4d03d.xml
        at org.opencastproject.workspace.impl.WorkspaceImpl.downloadIfNecessary(WorkspaceImpl.java:525) ~[?:?]
        at org.opencastproject.workspace.impl.WorkspaceImpl.lambda$downloadIfNecessary$2(WorkspaceImpl.java:537) ~[?:?]
        at org.opencastproject.util.IoSupport.locked(IoSupport.java:418) ~[?:?]
        at org.opencastproject.workspace.impl.WorkspaceImpl.get(WorkspaceImpl.java:387) ~[?:?]
        at org.opencastproject.workspace.impl.WorkspaceImpl.read(WorkspaceImpl.java:422) ~[?:?]
        at org.opencastproject.metadata.dublincore.DublinCoreUtil.loadDublinCore(DublinCoreUtil.java:75) ~[?:?]
        at org.opencastproject.metadata.dublincore.DublinCoreUtil.lambda$loadEpisodeDublinCore$0(DublinCoreUtil.java:63) ~[?:?]
        at java.util.Optional.map(Optional.java:260) ~[?:?]
        at org.opencastproject.metadata.dublincore.DublinCoreUtil.loadEpisodeDublinCore(DublinCoreUtil.java:63) ~[?:?]
        at org.opencastproject.search.impl.SearchServiceIndex.indexMediaPackage(SearchServiceIndex.java:285) ~[?:?]
        at org.opencastproject.search.impl.SearchServiceIndex.indexMediaPackage(SearchServiceIndex.java:275) ~[?:?]
        at org.opencastproject.search.impl.SearchServiceIndex.addSynchronously(SearchServiceIndex.java:254) ~[?:?]
        at org.opencastproject.search.impl.SearchServiceImpl.lambda$process$0(SearchServiceImpl.java:264) ~[?:?]
        at org.opencastproject.security.util.SecurityUtil.runAs(SecurityUtil.java:85) ~[?:?]
        at org.opencastproject.search.impl.SearchServiceImpl.process(SearchServiceImpl.java:262) ~[?:?]
        at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:314) ~[?:?]
        at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:272) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: org.opencastproject.security.api.TrustedHttpClientException: java.net.SocketTimeoutException: Read timed out
        at org.opencastproject.kernel.security.TrustedHttpClientImpl.execute(TrustedHttpClientImpl.java:423) ~[?:?]
        at org.opencastproject.kernel.security.TrustedHttpClientImpl.execute(TrustedHttpClientImpl.java:370) ~[?:?]
        at org.opencastproject.workspace.impl.WorkspaceImpl.downloadIfNecessary(WorkspaceImpl.java:503) ~[?:?]
        ... 20 more
Caused by: java.net.SocketTimeoutException: Read timed out
```

Which is *weird*, because `curl`-ing to localhost:8081 works fine.  I *suspect* here, that the CI tests use of an otherwise unpriviledged user (`ROLE_STUDIO` only) is partially to blame here, but I'm out of time to keep digging.

Arnei commented:
I also looked at this a couple weeks ago and came to the conclusion that the authorization is not working as it should, meaning I failed. I could not pinpoint exactly what I did wrong though.

gregorydlogan commented on modules/authorization-xacml/src/main/java/org/opencastproject/authorization/xacml/XACMLAuthorizationService.java:
```diff
@@ -308,44 +310,101 @@ private Optional<AccessControlList> loadAcl(final URI uri) {
-            return false;
+      if (entry.getAction().equals(action)) {
+        for (Role role : user.getRoles()) {
+          if (entry.getRole().equals(role.getName())) {
```
Leaving aside the rest of this PR, deny rules are not a thing in Opencast and will Soon(tm) go away via https://github.com/opencast/opencast/issues/7718#issuecomment-4782561978

gregorydlogan commented on modules/authorization-xacml/src/main/java/org/opencastproject/authorization/xacml/XACMLAuthorizationService.java:
```diff
@@ -308,44 +310,101 @@ private Optional<AccessControlList> loadAcl(final URI uri) {
       }
     }
-    logger.debug("XACML file allowed access");
+    logger.debug("ACL allowed access");
```
Since we're touching this line, can we add the MP id, and user?  Or just remove this line completely?
